samples/bluetooth/tmap_bmr/src/bap_broadcast_sink.c

```diff
@@ -0,0 +1,414 @@
+	if (err != 0) {
+		printk("Could not sync to PA: %d", err);
+
+	} else {
+		bcast_id = broadcast_id;
+	}
+}
+
+static bool scan_check_and_sync_broadcast(struct bt_data *data, void *user_data)
+{
+	uint32_t *broadcast_id = user_data;
+	struct bt_uuid_16 adv_uuid;
+
+	if (data->type != BT_DATA_SVC_DATA16) {
+		return true;
+	}
+
+	if (!bt_uuid_create(&adv_uuid.uuid, data->data, BT_UUID_SIZE_16)) {
+		return true;
+	}
+
+	if (bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_BROADCAST_AUDIO)) {
+		if (!bt_uuid_cmp(&adv_uuid.uuid, BT_UUID_TMAS)) {
+			struct net_buf_simple tmas_svc_data;
+			struct bt_uuid *uuid;
+			uint16_t uuid_val;
+			uint16_t peer_tmap_role = 0;
+
+			net_buf_simple_init_with_data(&tmas_svc_data,
+						      (void *)data->data,
+						      data->data_len);
+			uuid_val = net_buf_simple_pull_le16(&tmas_svc_data);
+			uuid = BT_UUID_DECLARE_16(sys_le16_to_cpu(uuid_val));
+			if (tmas_svc_data.len < sizeof(peer_tmap_role)) {
+				return false;
+			}
+			peer_tmap_role = net_buf_simple_pull_le16(&tmas_svc_data);
+			if ((sys_le16_to_cpu(peer_tmap_role) & BT_TMAP_ROLE_BMS)) {
+				printk("Found TMAP BMS\n");
+				tmap_bms_found = true;
+			}
+		}
+	} else {
+		*broadcast_id = sys_get_le24(data->data + BT_UUID_SIZE_16);
+	}
+
+	return true;
+}
+
+static void broadcast_scan_recv(const struct bt_le_scan_recv_info *info,
+				struct net_buf_simple *ad)
+{
+	struct net_buf_simple_state state;
+	uint32_t broadcast_id;
+
+	tmap_bms_found = false;
+
+	/* We are only interested in non-connectable periodic advertisers */
+	if ((info->adv_props & BT_GAP_ADV_PROP_CONNECTABLE) ||
+	     info->interval == 0) {
+		return;
+	}
+
+	net_buf_simple_save(ad, &state);
+	broadcast_id = INVALID_BROADCAST_ID;
+	bt_data_parse(ad, scan_check_and_sync_broadcast, (void *)&broadcast_id);
+	net_buf_simple_restore(ad, &state);
+
+	if ((broadcast_id != INVALID_BROADCAST_ID) && tmap_bms_found) {
+		sync_broadcast_pa(info, broadcast_id);
+	}
+}
+
+static void broadcast_scan_timeout(void)
+{
+	printk("Broadcast scan timed out\n");
+}
+
+static bool pa_decode_base(struct bt_data *data, void *user_data)
+{
+	struct bt_bap_base base = { 0 };
+	uint32_t base_bis_index_bitfield = 0U;
+
+	if (data->type != BT_DATA_SVC_DATA16) {
+		return true;
+	}
+
+	if (data->data_len < BT_BAP_BASE_MIN_SIZE) {
+		return true;
+	}
+
+	if (bt_bap_decode_base(data, &base) != 0) {
+		return false;
+	}
+
+	for (size_t i = 0U; i < base.subgroup_count; i++) {
+		for (size_t j = 0U; j < base.subgroups[i].bis_count; j++) {
+			const uint8_t index = base.subgroups[i].bis_data[j].index;
+
+			base_bis_index_bitfield |= BIT(index);
+		}
+	}
+
+	bis_index_bitfield = base_bis_index_bitfield & bis_index_mask;
+	k_sem_give(&sem_base_received);
+
+	return false;
+}
+
+static void broadcast_pa_recv(struct bt_le_per_adv_sync *sync,
+			      const struct bt_le_per_adv_sync_recv_info *info,
+			      struct net_buf_simple *buf)
+{
+	bt_data_parse(buf, pa_decode_base, NULL);
```

Comment:
Could consider just putting the content of `pa_decode_base` in here, instead of having this as a single-line function that calls another function